### PR TITLE
Consolidating by domain proposal

### PR DESCRIPTION
**Pull request by Numbers88s**
WIP

POC of how we could build our Graphql layer by modifying its relationship to the store.

### CODE REVIEW

File: cmd/frontend/services/executors/graphql/resolvers.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+func (e *ExecutorResolver) GitVersion() string      { return e.executor.GitVersion }
+func (e *ExecutorResolver) IgniteVersion() string   { return e.executor.IgniteVersion }
+func (e *ExecutorResolver) FirstSeenAt() DateTime {
+	return DateTime{Time: e.executor.FirstSeenAt}
+}
+func (e *ExecutorResolver) LastSeenAt() DateTime {
+	return DateTime{Time: e.executor.LastSeenAt}
+}
+
+// TODO: abstract this out into a common package
```
Comment (Numbers88s) [inline cmd/frontend/services/executors/graphql/resolvers.go:44]: Personal TODO

**Comment by efritz** [comment]
Could you give a high level sketch or dir/package structure?

**Comment by Numbers88s** [comment]
Definitely, here's a high level one:

architecture_diagram

To add context, we currently have one separation of concern/domain. Our current domain is "Code Intel". I am proposing keeping that domain, as we have other teams 😄, and adding a code intel specific separation of domain grouped by services. 

In other words separate Uploads, Index, Configuration Policies, and whatever else we grow into into their separate services. I think this would make it easier for us to continue to grow and maintain our backend.

The proposed architecture separates the functionality into three levels:
- **Transport Layer**:  Given that its a mostly GraphQL API this will be tied directly to the schema. However that is not always the case. Sometimes we also have http requests for a particular domain. Both can live at this level. Also separating it like this would allow for us to also perhaps integrate Protobuf (Thanks Olaf 😄 ). As a quick aside, I also like the fact that we can have validation at this level i.e. make sure the args passed are good to go or that the user has access, etc.
- **Service Layer**: Here is where I propose we have all the business logic and/or application logic. We can initialize our service with any clients it may need, logging, metrics, observability, etc, as well as its own store (relevant to its domain).  Perhaps some services need multiple types of store (LSIF & DB) so they would be attached to the service that needs it.
- **Store Layer**: This one is the easy one. All the db stuff, only this time its separated and contained within its domain.

That's the general idea. Let me know what you think. I've been iterating over this architecture pattern and its pretty easy to navigate once its in place but I am hoping to get more feedback ✌️

**Comment by Numbers88s** [comment]
This is the package architecture, in case that helps as well.

package-arch

**Comment by efritz** [comment]
FYI, reading through collapsed version: https://github.com/sourcegraph/sourcegraph/pull/28677

**Pull request closed without merging**